File: js/lib/schema.ts
import {
    ValueNotFound,
    ListConditionsNotFound,
    ListValueIsNotFound,
    ConditionsMissing,
    PrimitiveMissing,
    PrimitiveNotDefined
} from './error';
import { Primitives } from './primitives';

const KEYWORDS_GENERIC = Object.freeze({
    IS: ":is",
    KEY: ":key:",
    VALUE: ":value",
    CONDITIONS: ":conditions"
});

const KEYWORDS_LIST = Object.freeze({
    MIN_LENGTH: ":min_length",
    MAX_LENGTH: ":max_length"
});

const TYPE_CONTAINERS = Object.freeze({
    LIST: "list",
    OBJECT: "object"
});

function checkType(of: Readonly<any>, value: string) {
    return Object.keys(of).find(key => of[key] === value);
};

function isListOrObject(value: string) {
    return checkType(TYPE_CONTAINERS, value);
};

function checkIfList(value: string) {
    return value == TYPE_CONTAINERS.LIST;
};

function isKey(value: string) {
    return value.toString().includes(KEYWORDS_GENERIC.KEY) ?
        value.split(KEYWORDS_GENERIC.KEY).pop() : false;
};


class SchemaProcessor {
    private schema: Record<string, any> = {};

    private processList(schema: Record<string, any>, response: any[]): any[] {
        const _value = schema[KEYWORDS_GENERIC.VALUE];
        if (!schema[KEYWORDS_LIST.MIN_LENGTH] || !schema[KEYWORDS_LIST.MAX_LENGTH]) throw new ListConditionsNotFound(); 
        if (!_value[KEYWORDS_GENERIC.IS]) throw new ListValueIsNotFound();
        for (var _ in Array.from(Array(Primitives.integer({
            from: schema[KEYWORDS_LIST.MIN_LENGTH],
            to: schema[KEYWORDS_LIST.MAX_LENGTH]
        })).keys())) {
            response.push(this._process(_value));
        };
        return response;
    };

    private processObjecct(schema: Record<string, any>, response: Record<string, any[]>) {
        const _value = schema[KEYWORDS_GENERIC.VALUE];
        for (var key in _value) {
            const extractedKey = isKey(key);
            if (extractedKey) {
                const _processed = this._process(_value[key]);
                if ((!!response) && (response.constructor === Array)) (response[extractedKey] as Array<any>).push(_processed)
                else response[extractedKey] = _processed;
            };
        };
        return response;
    };


    private processPrimitive(type: string, conditions: any) {
        try { return Primitives[type](conditions); }
        catch (error) {
            if (error instanceof TypeError) {
                if (error.message.includes('is not a function')) {
                    throw new PrimitiveMissing(type);
                } else if (error.message.includes('Cannot read property')) {
                    throw new ConditionsMissing(type);
                }
            }
        }
    };

    private _process(schema: Record<string, any>) {
        const is = schema[KEYWORDS_GENERIC.IS];
        if (isListOrObject(is)) {
            if (!schema[KEYWORDS_GENERIC.VALUE]) throw new ValueNotFound();
            return checkIfList(is) ? this.processList(schema, []) : this.processObjecct(schema, {});
        } else return this.processPrimitive(is, schema[KEYWORDS_GENERIC.CONDITIONS]);
    };

    register(name: string, fn: Function) {
        if (!fn) throw new PrimitiveNotDefined(name);
        Primitives[name] = fn;
    }

    process = () => this._process(this.schema);

    setSchema = (schema: Record<string, any>) => this.schema = schema;

    deregister = (name: string) => delete Primitives[name];
};

export default SchemaProcessor;

File: dart/lib/schema.dart
// ignore_for_file: constant_identifier_names
import 'dart:math';

import 'package:eski/events-handler.dart';
import 'package:eski/primitives.dart';

// ignore: non_constant_identifier_names
final KEYWORD_SCHEMAKEY_MAP = {
  "IS": ":is",
  "KEY": ":key:",
  "VALUE": ":value",
  "CONDITIONS": ":conditions",
  "MIN_LENGTH": ":min_length",
  "MAX_LENGTH": ":max_length",
  "LIST": "list",
  "OBJECT": "object"
};

enum KEYWORDS {
  IS,
  KEY,
  VALUE,
  CONDITIONS,
  MIN_LENGTH,
  MAX_LENGTH,
  LIST,
  OBJECT
}

String? rawKey(KEYWORDS keyword) {
  return KEYWORD_SCHEMAKEY_MAP[keyword.name];
}

bool isListOrObject(String value) {
  return value == rawKey(KEYWORDS.LIST) || value == rawKey(KEYWORDS.OBJECT);
}

bool checkIfList(String value) {
  return value == rawKey(KEYWORDS.LIST);
}

String? getKey(String value) {
  return value.contains(rawKey(KEYWORDS.KEY)!)
      ? value.split(rawKey(KEYWORDS.KEY)!).last
      : null;
}

class SchemaProcessor {
  Map<String, dynamic> schema = {};
  SchemaProcessor({required this.schema});

  _process(String key, Map<String, dynamic> schema) {
    // ignore: non_constant_identifier_names
    var IS = schema[rawKey(KEYWORDS.IS)];
    if (IS == null) {
      EventsHandler.shoutIsNotProvided(key);
    } else {
      if (isListOrObject(IS)) {
        if (schema[rawKey(KEYWORDS.VALUE)] == null) {
          EventsHandler.shoutValueNotProvided(key);
        }
        return checkIfList(IS)
            ? _processList(key, schema, [])
            : _processObject(key, schema, {});
      } else {
        return _processPrimitive(IS, schema[rawKey(KEYWORDS.CONDITIONS)]);
      }
    }
  }

  process() {
    try {
      return _process('\$', schema);
    } catch (e) {
      if (e is! SchemaValidationException) rethrow;
    }
  }

  int between(int min, int max) {
    if (min == max) return min;
    return min + Random().nextInt(max - min);
  }

  _processList(String key, Map<String, dynamic> schema, List response) {
    var value = schema[rawKey(KEYWORDS.VALUE)];
    if (value == null) return response;
    if (schema[rawKey(KEYWORDS.MIN_LENGTH)] == null ||
        schema[rawKey(KEYWORDS.MAX_LENGTH)] == null) {
      EventsHandler.shoutListConditionsNotProvided(key);
    }
    if (value[rawKey(KEYWORDS.IS)] == null) {
      EventsHandler.shoutValueNotProvided(key);
    }

    for (int i = 0;
        i <
            between(schema[rawKey(KEYWORDS.MIN_LENGTH)],
                schema[rawKey(KEYWORDS.MAX_LENGTH)]);
        i++) {
      response.add(_process(key, value));
    }
    return response;
  }

  _processObject(
      String parentKey, Map<String, dynamic> schema, dynamic response) {
    Map<String, dynamic>? value = schema[rawKey(KEYWORDS.VALUE)];
    if (value == null) {
      EventsHandler.shoutValueNotProvided(parentKey);
    } else {
      for (String rawKey in value.keys) {
        String? key = getKey(rawKey);
        if (key != null) {
          response[key] = _process('$parentKey.$key', value[rawKey]);
        } else {
          EventsHandler.shoutKeyNotProvided(parentKey);
        }
      }
    }
    return response;
  }

  Map<Symbol, dynamic> _symbolizeKeys(Map<String, dynamic> map) {
    return map.map((k, v) => MapEntry(Symbol(k), v));
  }

  _processPrimitive(String type, dynamic conditions) {
    var primitive = Primitives[type];
    if (primitive == null) {
      EventsHandler.shoutPrimitiveNotAvailable(type);
    }
    try {
      return Function.apply(primitive!, [], _symbolizeKeys(conditions ?? {}));
    } catch (e) {
      if (e.toString().contains('Closure call with mismatched arguments')) {
        EventsHandler.shoutInvalidConditionsForPrimitive(type, conditions);
      }
      rethrow;
    }
  }
}
